zerver/lib: Fix annotations in cache.py and cache_helpers.py

sharmaeklavya2 opened the pull request:
When I tried to annotate fields in our models, I got a lot of mypy errors. Many annotations in zerver/lib/cache.py, zerver/lib/cache_helpers.py and zerver/models.py were wrong. So I first fixed them. I have made this PR now so that that work can be reviewed.

I have also added a `zerver/lib/text_utils.py`. This is inspired from [django.utils.encoding](https://docs.djangoproject.com/en/1.9/ref/utils/#module-django.utils.encoding).

Some of the intermediate commits don't pass mypy checks, but I couldn't figure out a way to split my commits otherwise. All commits pass tools/test-backend.

zerver/lib/cache.py had a lot of wrong annotations. I fixed as many as I could find. But that produced other mypy errors. When I tried to fix them, I got other mypy errors. This chain continued for some time until I hit a few places where `bytes` were being used (like zlib.compress and decompress in zerver/models.py). Those had bugs which prevented me from annotating some functions correctly, so I had to fix them.

timabbott commented:
Cool, thanks for working on this @sharmaeklavya2!  Just to confirm I understand `force_str` correctly, is the idea that it will do the conversation from str/unicode (in python 2) or bytes/str (in python 3) to end up with a str regardless of whether the code is running on python 2 or 3?

timabbott commented:
I merged `zerver/lib/cache.py: Remove unneeded return statements.`, since that commit looked good; still reviewing the rest.

timabbott commented on zerver/models.py:
```diff
@@ -748,20 +751,20 @@ def linebreak(string):
-    # type: (Dict[Any, Any]) -> str
-    return zlib.compress(ujson.dumps(message_dict).encode("utf-8"))
+    # type: (Dict[Any, Any]) -> binary_type
+    return zlib.compress(force_bytes(ujson.dumps(message_dict)))
```
`stringify_message_dict` and `extract_message_dict` should be inverses of each other; maybe that should be reflected in the type of the input to stringify_message_dict?

sharmaeklavya2 replied:
Oh, I missed that. You're right. The parameter type should be `Dict[text_type, Any]` instead of `Dict[Any, Any]`.

sharmaeklavya2 commented:
Yes, `force_str` will return an `str` on both python 2 and python 3. So on python 2 if I pass a `unicode` to `force_str`, it will encode it to return an `str`. On python 3 if I pass `bytes` to it it will decode it to return `str`.

timabbott commented:
OK, that makes sense.  Structurally, the big thing here is that this branch converts the cache keys for memcached to by unicode; I think we should check carefully whether `memcached` (or the Django interface to it that we're using) actually supports unicode keys; it's possible we should instead be making our cache keys `binary_type`.  Can you investigate this @sharmaeklavya2?

sharmaeklavya2 commented:
Yeah I had investigated this. memcached actually converts its keys to `unicode` on python 2.

timabbott commented:
OK, great, thanks for checking that!  A few more comments:
* zerver/lib/text_utils.py deserves some comments/docstrings explaining roughly what we discussed (at least for force_str, and it wouldn't hurt to have something for the rest about when these functions should be used). 
* I think it'd be good to, in the last second-to-last-commit, split the annotation errors that are unrelated to cache keys moving from str=>text_type into a separate commit or commits (e.g. the UserProfile/Realm stuff in zerver/lib/cache.py).

timabbott commented on zerver/lib/cache.py:
```diff
@@ -46,16 +54,17 @@ def remote_cache_stats_finish():
         # This sets the prefix mostly for the benefit of the JS tests.
         # The Python tests overwrite KEY_PREFIX on each test.
-        return 'test_suite:' + str(os.getpid()) + ':'
+        return 'test_suite:' + text_type(os.getpid()) + ':'
```
Hmm, this is a bit confusing, because the LHS and RHS of this computation are `str` still; maybe clearest would be `u"test_suite:%s:" % (os.getpid(),)` ?

sharmaeklavya2 replied:
Yeah I agree. I'll change that.

timabbott commented on zerver/lib/cache.py:
```diff
@@ -77,15 +86,15 @@ def get_or_create_key_prefix():
+    # type: (text_type) -> None
     global KEY_PREFIX
-    KEY_PREFIX = test_name + ':' + str(os.getpid()) + ':'
+    KEY_PREFIX = test_name + ':' + text_type(os.getpid()) + ':'
```
same comment as above on KEY_PREFIX

sharmaeklavya2 commented:
Here is an example which proves that memcached converts keys to unicode.

```
InteractiveConsole)
>>> from zerver.lib.cache import get_cache_backend
>>> memcached = get_cache_backend('django.core.cache.backends.memcached.PyLibMCCache')
>>> memcached
<django.core.cache.backends.memcached.PyLibMCCache object at 0x7fd85080e150>
>>>
>>> memcached.set(1, 'a')
>>> memcached.get(1)
'a'
>>> memcached.get('1')
'a'
>>> memcached.get(u'1')
'a'
>>>
>>> smiley = u'\U0001f600'
>>> bsmiley = smiley.encode('utf-8')
>>> bsmiley
'\xf0\x9f\x98\x80'
>>> print(smiley)
😀
>>> print(bsmiley)
😀
>>> memcached.set(smiley, 'smile!')
>>> memcached.get(smiley)
'smile!'
>>> memcached.get(bsmiley)
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/mnt/Code/git_repos/forked/zvenv2/local/lib/python2.7/site-packages/django/core/cache/backends/memcached.py", line 83, in get
    key = self.make_key(key, version=version)
  File "/mnt/Code/git_repos/forked/zvenv2/local/lib/python2.7/site-packages/django/core/cache/backends/memcached.py", line 76, in make_key
    return force_str(super(BaseMemcachedCache, self).make_key(key, version))
  File "/mnt/Code/git_repos/forked/zvenv2/local/lib/python2.7/site-packages/django/core/cache/backends/base.py", line 102, in make_key
    new_key = self.key_func(key, self.key_prefix, version)
  File "/mnt/Code/git_repos/forked/zvenv2/local/lib/python2.7/site-packages/django/core/cache/backends/base.py", line 35, in default_key_func
    return '%s:%s:%s' % (key_prefix, version, key)
UnicodeDecodeError: 'ascii' codec can't decode byte 0xf0 in position 0: ordinal not in range(128)
>>> memcached.delete(smiley)
>>> memcached.set(bsmiley, 'smile!')
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/mnt/Code/git_repos/forked/zvenv2/local/lib/python2.7/site-packages/django/core/cache/backends/memcached.py", line 90, in set
    key = self.make_key(key, version=version)
  File "/mnt/Code/git_repos/forked/zvenv2/local/lib/python2.7/site-packages/django/core/cache/backends/memcached.py", line 76, in make_key
    return force_str(super(BaseMemcachedCache, self).make_key(key, version))
  File "/mnt/Code/git_repos/forked/zvenv2/local/lib/python2.7/site-packages/django/core/cache/backends/base.py", line 102, in make_key
    new_key = self.key_func(key, self.key_prefix, version)
  File "/mnt/Code/git_repos/forked/zvenv2/local/lib/python2.7/site-packages/django/core/cache/backends/base.py", line 35, in default_key_func
    return '%s:%s:%s' % (key_prefix, version, key)
UnicodeDecodeError: 'ascii' codec can't decode byte 0xf0 in position 0: ordinal not in range(128)
>>>
```

sharmaeklavya2 commented:
Github is showing commits in the wrong order. Here is the output from `git log --pretty=oneline --abbrev-commit`:

```
b794ad1 Fix annotations incompatible with zerver/lib/cache.py.
8db3057 Fix annotations in zerver/lib/cache.py and cache_helpers.py.
e8d8150 Use text_type as type of cache_keys.
1b920da Make encoding/decoding strings python 3 compatible.
c1a38bf Add zerver/lib/text_utils.py.
```

timabbott commented:
Cool, sounds good; looks like you haven't processed all my comments yet; let me know when you have!

sharmaeklavya2 commented:
@timabbott, I have made the changes you suggested. I have decided that I'll annotate model fields in a separate pull request, so I changed the pull request title.

timabbott commented:
Cool, thanks @sharmaeklavya2!  Annotating model fields in another PR makes sense, since this is somewhat involved already.

It looks like you left off adding comments/docstrings in zerver/lib/text_utils.py -- can you add those?  Just FYI, here's how I would probably have structured this branch to not have any failing checks on any of the individual commits:

* First, the commit adding text_utils.py
* `Use text_type as type of cache_keys.`, `zerver/lib: Fix annotations in cache.py and cache_helpers.py.`, and `Fix annotations incompatible with zerver/lib/cache.py.` would be mostly squashed into one commit that is basically "Use text_type"; the things I would pull out are the changes that are fundamentally unrelated (e.g. adding TypeVars for generic_bulk_cached_fetch should be its own commit, as should fixing the None return value of `update_user_profile_caches` in "zerver/lib: Fix annotations in cache.py and cache_helpers.py.").  The overall result would be that the things which aren't fixing str->text_type would all be their own commits, and then there'd be one coherent commit which switches cache keys from str->text_type in the annotations and propagates the types into the ~5 files that use those types.
* Last, the commit replacing Any with UserProfile/Realm in the cache system.  This change doesn't need to be tangled with the text_type changes.

I think this would be a fairly easy rebase to do, so if you're up for it, it'd be great to restructure the branch that way, but what's more important from my perspective is that you understand the technique around avoiding getting into this sort of situation in the first place.  Ideally, while trying to propagate the str->text_type changes, when you noticed that we have all these `Any` values that should be a UserProfile or Realm in that same file, I hope you'll get into the habit of either:
(1) write a note to yourself to fix that once you get the str->text_type migration fully passing, and then do it as an extra commit at the end of the migration, or
(2) committing or stashing your partially finished str->text_type migration changes, doing a `git rebase -i`, and adding a commit at the start of your branch to fix the types (and get that fully passing checks), and then proceed with the migration.

At least, that's how I usually do things, and I find it works pretty well for avoiding ending up with a branch that's broken for a while in the middle.

(Also, one more sidenote: The `if False` stuff is not going to work when we move to the Python 3 type syntax, so we will eventually want to solve those import loops another way, which probably involves splitting zerver/lib/cache.py into multiple files)

sharmaeklavya2 commented:
If 7999751 (`Use text_type as type of cache_keys.`), 0a71a61 (`zerver/lib: Fix annotations in cache.py and cache_helpers.py.`) and 0763a21(`Fix annotations incompatible with zerver/lib/cache.py.`) are squashed together, all commits of my PR will pass mypy checks. All these 3 commits only involve migrating from `str` to `text_type`. 

I had to add 2 ignore statements in d3b80d9 (`Use appropriate string types and correctly encode/decode them.`) to make it pass mypy checks. I later removed those ignore statements in 0a71a61 (`zerver/lib: Fix annotations in cache.py and cache_helpers.py.`)

I have renamed `text_utils.py` to `str_utils.py`, because this file is also meant for bytestrings which are not text. I have also added documentation about the `force` functions as docstrings and referred to them from docs/mypy.md.

timabbott commented:
Awesome, this looks great @sharmaeklavya2!  I went ahead and squashed those 3 commits together and merged this branch.  Thanks for cleaning this up, I really appreciate it.
